src/C++/SSLSocketInitiator.cpp

```diff
@@ -318,83 +323,163 @@ void SSLSocketInitiator::doConnect( const SessionID& s, const Dictionary& d )
     setPending( s );
     m_pendingConnections[ result ] = new SSLSocketConnection( *this, s, result, ssl, &m_connector.getMonitor() );
   }
-  catch ( std::exception& ) {}
+  catch ( std::exception& ex)
+  {
+      getLog()->onEvent(ex.what());
+  }
 }
 
-bool SSLSocketInitiator::handshakeSSL(SSL* ssl)
+SSLHandshakeStatus SSLSocketInitiator::handshakeSSL(SSLSocketConnection* connection)
 {
+    SSL* ssl = connection->sslObject();
     ERR_clear_error();
     // Do the SSL handshake.
     int rc = SSL_connect(ssl);
-    while (rc <= 0) {
+    if (rc <= 0) {
         int err = SSL_get_error(ssl, rc);
         if ((err == SSL_ERROR_WANT_READ) ||
             (err == SSL_ERROR_WANT_WRITE)) {
             errno = EINTR;
-        } else {
+
+            if (err == SSL_ERROR_WANT_WRITE) {
+                connection->subscribeToSocketWriteAvailableEvents();                
+            }
+            return SSL_HANDSHAKE_IN_PROGRESS;
+        }
+        else if (err == SSL_ERROR_SYSCALL)
+        {
+            getLog()->onEvent("SSL_connect failed with SSL error " + IntConvertor::convert(err) + ". Error stack:");
+
+            char errorBuffer[512];
+            unsigned long systemError;
+
+            while ((systemError = ERR_get_error()) != 0) {
+                ERR_error_string_n(systemError, errorBuffer, sizeof(errorBuffer));
+                getLog()->onEvent(errorBuffer);
+            }
+            getLog()->onEvent("End of error stack");
+
+            getLog()->onEvent(socket_get_last_error());
+
+            return SSL_HANDSHAKE_FAILED;
+        }
+        else {            
             getLog()->onEvent("SSL_connect failed with SSL error " + IntConvertor::convert(err));
-            return false;
+            return SSL_HANDSHAKE_FAILED;
         }
-        ERR_clear_error();
-        rc = SSL_connect(ssl);
     }
-    return true;
+
+    return SSL_HANDSHAKE_SUCCEDED;
 }
 
 void SSLSocketInitiator::onConnect( SocketConnector& connector, socket_handle s )
 {
-  SocketConnections::iterator i = m_pendingConnections.find( s );
-  if( i == m_pendingConnections.end() ) return;
+  getLog()->onEvent("Socket connected handle: " + std::to_string(s));
+   
+  time_t now;
+  ::time(&now);
+
+  SocketConnections::iterator i = m_pendingConnections.find(s);
+  if (i == m_pendingConnections.end()) return;
   SSLSocketConnection* pSocketConnection = i->second;
+  
```

Comment:
Below the connection is removed from the pendingConnections list, and is placed into the pendingSSLHandshakes list.